sklearn/_build_utils/__init__.py

```diff
@@ -75,3 +75,26 @@ def maybe_cythonize_extensions(top_path, config):
             nthreads=n_jobs,
             compile_time_env={'SKLEARN_OPENMP_SUPPORTED': with_openmp},
             compiler_directives={'language_level': 3})
+
+
+def gen_from_templates(templates):
+    """Generate cython files from templates"""
+    if not isinstance(templates, list):
+        templates = [templates]
+
+    for template in templates:
+        pyxfile = template.replace('.tp', '')
+
+        if not (os.path.exists(pyxfile) and
+                os.stat(template).st_mtime < os.stat(pyxfile).st_mtime):
+
+            with open(template, "r") as f:
+                tmpl = f.read()
+
+            # Lazy import because cython is not a dependency when building from
+            # source distribution.
+            from Cython import Tempita # noqa
```

Comment:
Put this at the top of the function?

Reply:
The goal is a lazy import of tempita. We need to import it only if we are going to transform the template. For example when building from source distribution, we ship already cythonized files but we still run the setup to build from c and c++ files.

Reply:
Yes, but 

- Pretty sure the `if` clause above is always True in all reasonable scenarios
- Pretty sure we should remove it anyway

I propose to put the import at the top of the function, not top of the module.  The import is still lazy

Reply:
this function is called in the setup function which is called when sklearn is built from source distribution. In that case cython is not a dependency. So we can't put the import at the top of the function. When building from source distribution the cythonize part has already been done so we don't enter the if clause. This is why we need the import at this moment not before.

Reply:
Also we don't have a cython runtime dependency. Moving the import at the top of the function would make the tests testing the setup functions fail.

Reply:
Sorry, not following.

As far as I understand the only scenario where Cython may not be required for building is when we ship the .c files. This PR doesn't check for the existence of .c files but .py[xd] files.

Unless you disagree with me that the `if` clause above is/should always be True (I might be wrong!), there is no point in importing Cython here.

Reply:
> This PR 

note that this `if` condition is not introduced in this PR. It was there and I'm just moving it in a utility function.

> This PR doesn't check for the existence of .c files but .py[xd] 

I agree that the condition seems odd but it "works" in the sense that that before building you have only .pyx.tp files and then you have .pyx and .c files so checking the existence of .pyx is equivalent to checking the existence of .c files.

I agree that we probably improve this condition but I'd like to have the opinion of the ones who wrote it first to make sure we are not missing something.